### DISCUSSION TITLE
having trouble getting DPS1 to work consistently with low power library

### DISCUSSION THREAD

**Discussion by whyameye** [dbuezas/lgt8fx#316] [Q&A]
I have great luck using the low power library when going into DPS2 but have been having trouble getting consistent and understandable behavior with other low power modes. Below is an example with comments showing how DPS1 is not consistently giving me the results I expect. Likely I'm misunderstanding the [documentation in the low power library](https://github.com/dbuezas/lgt8fx/tree/master/lgt8f/libraries/lgt_LowPower) and [the uC datasheet](https://github.com/dbuezas/lgt8fx/blob/master/docs/LGT8FX8P_databook_v1.0.5-English.pdf) but haven't been making much progress on my own:

```
// circuit: D2 or D3 connected to a pushbutton which, when pressed, connects to GND
// pin has external 10K pullup (would internal pullup work as well? Seemed inconsistent when I tried.)
// (how to detect changes on other port D pins such as D7? Changing PD_PIN to 7 and moving the pushbutton to pin D7 does not work)

#define PD_PIN 3

#include "lgt_LowPower.h"
volatile boolean interrupt = false;

// without this routine and matching attachInterrupt uC doesn't wake up on button release. Why?
void someISR() {
  interrupt = true;
  return;
}

void setup() {
  Serial.begin(115200);
  delay(100);
  Serial.println("starting");
  pinMode(PD_PIN, INPUT);
  delay(500);
  // won't wake up without the below line. Why?
  attachInterrupt(digitalPinToInterrupt(PD_PIN), someISR, HIGH);
}

void loop() {
  while (digitalRead(PD_PIN)) {
    Serial.print("D");
    Serial.print(PD_PIN);
    Serial.println(" high");
    delay(500);
  }
  Serial.print("D");
  Serial.print(PD_PIN);
  Serial.println(" low. Going to sleep");
  delay(100);

  // is this necessary for DPS1? Sometimes uC wakes up and sometimes doesn't when this is commented out:
  IOCWK = (1 << PD_PIN);

  // wakes up from releasing the pushbutton MOST of the time but not always even with IOCWK set
  // sometimes also resets instead of continuing to execute the code
  // also usually does not wake up at all regardless of button state if SLEEP_32S is changed to SLEEP_FOREVER (but sometimes does wake up)
  LowPower.powerDown(SLEEP_32S, ADC_OFF, BOD_OFF);
  delay(100);
  if (interrupt) {
    Serial.println("interrupt!");
    interrupt = false;
  }
  Serial.println("waking up");
  pinMode(PD_PIN, INPUT);
}
```

**Comment by whyameye** [comment]
@LaZsolt I was wondering if you might have any insight on this you could share with me.

**Comment by LaZsolt** [comment]
Hello John!

Please excuse me, I will try to be brief.

**1.**
```
 // won't wake up without the below line. Why?
  attachInterrupt(digitalPinToInterrupt(PD_PIN), someISR, HIGH);
```

LGT8F328P will wake-up only when enabling interrupt. This behaviour is different from ATmega32P

**2.**
I think the IOCWK register should only be used in DPS2 mode.

**3.**
Use `Serial.flush();` instead of `delay();` to wait for the end of serial data transfer before sleep.


Watch this example: https://github.com/dbuezas/lgt8fx/blob/master/lgt8f/libraries/lgt_LowPower/examples/standbyExternalInterrupt/standbyExternalInterrupt.ino

**4.**
Important to set up the I/O pins for low consumption before sleep. It is recommended to set all input pins with internal pull up resistors.
The wires which connected to the (input) pins, without pull up resistors, can easily generate noises which may cause wake-up event.

**Comment by whyameye** [comment]
I'm glad you suggested this example:
https://github.com/dbuezas/lgt8fx/blob/master/lgt8f/libraries/lgt_LowPower/examples/standbyExternalInterrupt/standbyExternalInterrupt.ino

because it exactly shows the erratic behavior I'm talking about. I ran this multiple times, sometimes by resetting the MCU  and sometimes by reprogramming. Every time I ran this I got different incorrect results.

Here's one example output I get running the above linked code. I made no change to the code. D2 is connected to a button that when pressed connects to GND:
```
17:38:42.458 -> 
17:38:42.458 -> Sleep and external interrupt example started.
17:38:42.562 -> 
17:38:42.562 -> Entering standby mode.
17:38:42.563 -> Apply low signal to wake the MCU.
I pressed the button here and got the correct behavior this time:
17:38:45.291 -> Awake!
17:38:45.291 -> MCU was woken up via an external interrupt pin.
17:38:49.288 -> 
17:38:49.288 -> Entering standby mode.
17:38:49.288 -> Apply low signal to wake the MCU.
I pressed the button but instead of continuing to execute the code the MCU reset itself:
17:38:50.128 -> 
17:38:52.769 -> Sleep and external interrupt example started.
17:38:52.834 -> 
17:38:52.834 -> Entering standby mode.
17:38:52.897 -> Apply low signal to wake the MCU.
not sure whether I pressed the button or not here but either way it's clear the resulting behavior is incorrect
17:38:55.827 -> H�����external interrupt example started.
17:38:55.859 -> 
```

Thanks!

John

**Reply by LaZsolt** [reply]
Sorry, I have no more ideas atm.

**Comment by whyameye** [comment]
since the example code doesn't work correctly I will submit as an issue then.
